Source/Core/Common/Subscribable.h

```diff
@@ -0,0 +1,78 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <map>
+#include <memory>
+
+#include "Common/NoDiscard.h"
+#include "Common/NonCopyable.h"
+
+template <typename... Args>
+class Subscribable : NonCopyable
+{
+public:
+  using SubscriptionID = uint32_t;
+  using CallbackType =
+      std::function<void(std::add_lvalue_reference_t<std::add_const_t<std::decay_t<Args>>>...)>;
+
+  class Subscription
+  {
+    friend class Subscribable;
+
+  public:
+    Subscription() = default;
```

Comment:
Yeah, this is here to allow default initialization, so you don't have to assign to a variable immediately when you create it. The STL's `unique_ptr` and `shared_ptr` objects also allow this, because it's useful for constructors (you don't have to assign inside the member initializers list if you don't want to) and other situations where declaration and initialization aren't in the same location.

Here's an example from the test file that uses default initialization:

```c++
std::array<int, 10> values{};
std::array<Subscribable<int>::Subscription, 10> subscriptions;
for (size_t i = 0; i < values.size(); ++i)
{
  subscriptions[i] = s->Subscribe([i, &values](int new_value) { values[i] = new_value; });
}
```

Without a public default initializer, this would have to look more like:

```c++
std::array<int, 10> values{};
std::array<Subscribable<int>::Subscription, 10> subscriptions = {{
    s->Subscribe([&values](int new_value) { values[0] = new_value; }),
    s->Subscribe([&values](int new_value) { values[1] = new_value; }),
    s->Subscribe([&values](int new_value) { values[2] = new_value; }),
    // ...
    s->Subscribe([&values](int new_value) { values[9] = new_value; })
}};
```

Yuck!